elixir/apps/domain/lib/domain/events/event.ex

```diff
@@ -16,11 +16,49 @@ defmodule Domain.Events.Event do
     old_data = zip(old_tuple_data, relation.columns)
     data = zip(tuple_data, relation.columns)
 
+    log(op, table, old_data, data)
```

Comment:
I want to raise attention to the order here. This means we insert the log entry and _then_ handle side effects.

We could also do it the other way around. The inserts will be very fast as they don't check for locks or anything.

The important thing is that the entire ingest pipeline here is single-threaded so that we only respond with success back to Postgres if we've handled all of things we need to.

If we crash or something while handling these, we won't respond, and postgres will resend the message the next time we connect.

Reply:
I also want to call attention to the following minor challenge with this design.

We can speed these inserts up by a lot if we batch them with `insert_all/3`. However, then we break the single-threaded nature of this thing, since we would have to queue the insert instead of write it. Presumably we could queue messages, then acknowledge them, then fail to `insert_all/3` them, which would result in data not making it to the `change_logs` table.

But, we have a solution. We can ACK to Postgres after the `insert_all/3` completes, with the LSN of the last message in the buffer. We'd also want to set a timeout here to avoid Postgres cutting our connection due to us not ACKing in a timely manner (15s I believe).

Reply:
My current thinking is to get this out and soak test it for a bit. Monitor resource usage closely on the domain node.

If it turns out to be nothing burger, we can kick this optimization much further down the road.

The risk here should be fairly low. This is a good place actually to add a trace to monitor our queue delay time.

Will push up a commit to do that.

Reply:
This discussion is largely nullified by splitting the change log consumer into its own replication connection.